# Business Problem

Human bidders on the site are becoming increasingly frustrated with their inability to win auctions vs. their software-controlled counterparts. As a result, usage from the site's core customer base is plummeting.

In order to rebuild customer happiness, the site owners need to eliminate computer generated bidding from their auctions. Their attempt at building a model to identify these bids using behavioral data, including bid frequency over short periods of time, has proven insufficient.

# Goal

Identify online auction bids that are placed by "robots", helping the site owners easily flag these users for removal from their site to prevent unfair auction activity. 

# Setup

In [3]:
import os
import numpy as np
import pandas as pd

# For data visualisation
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

# Where to save the figures
PROJECT_ROOT_DIR = "."
IMAGES_PATH = os.path.join(PROJECT_ROOT_DIR, "images")
os.makedirs(IMAGES_PATH, exist_ok=True)

def save_fig(fig_id, tight_layout=True, fig_extension="png", resolution=300):
    path = os.path.join(IMAGES_PATH, fig_id + "." + fig_extension)
    print("Saving figure", fig_id)
    if tight_layout:
        plt.tight_layout()
    plt.savefig(path, format=fig_extension, dpi=resolution)


# Get Data

In [4]:
bids = pd.read_csv("data/bids.csv")
bidders_train = pd.read_csv("data/train.csv")
bidders_test = pd.read_csv("data/test.csv")

## Understanding the data

**For the bidder dataset**
* bidder_id – Unique identifier of a bidder.
* payment_account – Payment account associated with a bidder. These are obfuscated to protect privacy. 
* address – Mailing address of a bidder. These are obfuscated to protect privacy. 
* outcome – Label of a bidder indicating whether or not it is a robot. Value 1.0 indicates a robot, where value 0.0 indicates human. 
    * The outcome was half hand labeled, half stats-based. There are two types of "bots" with different levels of proof:
        1. Bidders who are identified as bots/fraudulent with clear proof. Their accounts were banned by the auction site.
        2. Bidder who may have just started their business/clicks or their stats exceed from system wide average. There are no clear proof that they are bots. 


**For the bid dataset**
* bid_id - unique id for this bid
* bidder_id – Unique identifier of a bidder (same as the bidder_id used in train.csv and test.csv)
* auction – Unique identifier of an auction
* merchandise –  The category of the auction site campaign, which means the bidder might come to this site by way of searching for "home goods" but ended up bidding for "sporting goods" - and that leads to this field being "home goods". This categorical field could be a search term, or online advertisement. 
* device – Phone model of a visitor
* time - Time that the bid is made (transformed to protect privacy).
* country - The country that the IP belongs to
* ip – IP address of a bidder (obfuscated to protect privacy).
* url - url where the bidder was referred from (obfuscated to protect privacy). 

## Take a Quick Look at the Data Structure

In [5]:
# Get a sense of our data
bids.head()

bid_id                              bidder_id auction merchandise  device  \
0       0  8dac2b259fd1c6d1120e519fb1ac14fbqvax8   ewmzr     jewelry  phone0   
1       1  668d393e858e8126275433046bbd35c6tywop   aeqok   furniture  phone1   
2       2  aa5f360084278b35d746fa6af3a7a1a5ra3xe   wa00e  home goods  phone2   
3       3  3939ac3ef7d472a59a9c5f893dd3e39fh9ofi   jefix     jewelry  phone4   
4       4  8393c48eaf4b8fa96886edc7cf27b372dsibi   jefix     jewelry  phone5   

               time country              ip              url  
0  9759243157894736      us   69.166.231.58  vasstdc27m7nks3  
1  9759243157894736      in   50.201.125.84  jmqlhflrzwuay9c  
2  9759243157894736      py  112.54.208.157  vasstdc27m7nks3  
3  9759243157894736      in   18.99.175.133  vasstdc27m7nks3  
4  9759243157894736      in    145.138.5.37  vasstdc27m7nks3

In [23]:
# Check data types
bids.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7656334 entries, 0 to 7656333
Data columns (total 9 columns):
 #   Column       Dtype 
---  ------       ----- 
 0   bid_id       int64 
 1   bidder_id    object
 2   auction      object
 3   merchandise  object
 4   device       object
 5   time         int64 
 6   country      object
 7   ip           object
 8   url          object
dtypes: int64(2), object(7)
memory usage: 525.7+ MB


In [11]:
# Check for missing values
bids.isnull().sum()

bid_id            0
bidder_id         0
auction           0
merchandise       0
device            0
time              0
country        8859
ip                0
url               0
dtype: int64

In [20]:
# Check number of unique values in non-numerical attributes
bids.describe(include=['O'])

bidder_id  auction merchandise   device  \
count                                 7656334  7656334     7656334  7656334   
unique                                   6614    15051          10     7351   
top     f5b2bbad20d1d7ded3ed960393bec0f40u6hn    jqx39      mobile   phone4   
freq                                   515033   537347     2126587   706573   

        country             ip              url  
count   7647475        7656334          7656334  
unique      199        2303991          1786351  
top          in  37.95.134.218  vasstdc27m7nks3  
freq    1788731          61203          3499409

In [6]:
bidders_train.head()

bidder_id  \
0  91a3c57b13234af24875c56fb7e2b2f4rb56a   
1  624f258b49e77713fc34034560f93fb3hu3jo   
2  1c5f4fc669099bfbfac515cd26997bd12ruaj   
3  4bee9aba2abda51bf43d639013d6efe12iycd   
4  4ab12bc61c82ddd9c2d65e60555808acqgos1   

                         payment_account  \
0  a3d2de7675556553a5f08e4c88d2c228754av   
1  a3d2de7675556553a5f08e4c88d2c228v1sga   
2  a3d2de7675556553a5f08e4c88d2c2280cybl   
3  51d80e233f7b6a7dfdee484a3c120f3b2ita8   
4  a3d2de7675556553a5f08e4c88d2c22857ddh   

                                 address  outcome  
0  a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
1  ae87054e5a97a8f840a3991d12611fdcrfbq3      0.0  
2  92520288b50f03907041887884ba49c0cl0pd      0.0  
3  4cb9717c8ad7e88a9a284989dd79b98dbevyi      0.0  
4  2a96c3ce94b3be921e0296097b88b56a7x1ji      0.0

In [22]:
bidders_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013 entries, 0 to 2012
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   bidder_id        2013 non-null   object 
 1   payment_account  2013 non-null   object 
 2   address          2013 non-null   object 
 3   outcome          2013 non-null   float64
dtypes: float64(1), object(3)
memory usage: 63.0+ KB


In [7]:
bidders_test.head()

bidder_id  \
0  49bb5a3c944b8fc337981cc7a9ccae41u31d7   
1  a921612b85a1494456e74c09393ccb65ylp4y   
2  6b601e72a4d264dab9ace9d7b229b47479v6i   
3  eaf0ed0afc9689779417274b4791726cn5udi   
4  cdecd8d02ed8c6037e38042c7745f688mx5sf   

                         payment_account  \
0  a3d2de7675556553a5f08e4c88d2c228htx90   
1  a3d2de7675556553a5f08e4c88d2c228rs17i   
2  925381cce086b8cc9594eee1c77edf665zjpl   
3  a3d2de7675556553a5f08e4c88d2c228nclv5   
4  a3d2de7675556553a5f08e4c88d2c228dtdkd   

                                 address  
0  5d9fa1b71f992e7c7a106ce4b07a0a754le7c  
1  a3d2de7675556553a5f08e4c88d2c228klidn  
2  a3d2de7675556553a5f08e4c88d2c228aght0  
3  b5714de1fd69d4a0d2e39d59e53fe9e15vwat  
4  c3b363a3c3b838d58c85acf0fc9964cb4pnfa

* "country" have 8859 (0.1%) missing values. Considering the small proportion, let's ignore it first.
* "bid_id" current data type is int64, which should had been object instead since it is an unique identifier for each bid.
* "time" has been transformed to preserve privacy. Let's ignore it first in our baseline model.
* Currently our bids dataset and our target label is in separate datasets, we can't do much without the target label. Let's aggregate our bids dataset and merge it with the bidder_train set for further data exploration.

# Explore the Data

# Prepare the Data

# Shortlist Promising Models

# Fine-Tune the System